persist/fs/index_lookup.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/m3db/m3db/digest"
+	"github.com/m3db/m3db/persist/encoding"
+	"github.com/m3db/m3db/ts"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// nearestIndexOffsetLookup provides a way of quickly determining the nearest offset of an
+// ID in the index file. It is not safe for concurrent use
+type nearestIndexOffsetLookup struct {
+	summaryIDsOffsets []encoding.IndexSummaryToken
+	// bytes from file mmap'd into anonymous region
+	summariesBytes []byte
+	// reusable decoder stream
+	decoderStream  encoding.DecoderStream
+	decoder        encoding.Decoder
+	msgpackDecoder *msgpack.Decoder
+}
+
+func newNearestIndexOffsetLookup(
+	summaryIDsOffsets []encoding.IndexSummaryToken,
+	summariesBytes []byte,
+	decoder encoding.Decoder,
+	decoderStream encoding.DecoderStream,
+) *nearestIndexOffsetLookup {
+	return &nearestIndexOffsetLookup{
+		summaryIDsOffsets: summaryIDsOffsets,
+		summariesBytes:    summariesBytes,
```

Comment:
Considering we munmap with close, perhaps better to name this byte buffer so it's explicit and not weird when reading the code?  i.e. `summariesMmap`?

Reply:
sure